### PR TITLE
Patch 2

### PR DESCRIPTION
**Pull request by braingram**
This PR:
- adds a unit test for `weighted_operations` (see inline comment)
- cleans up `**kwargs` in `FringeFitter`

**Checklist for maintainers**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/ami/lg_model.py [hunk 1]
```diff
@@ -388,14 +388,14 @@ def fit_image(
             )
         else:
             self.fittingmodel = modelin
-        if self.weighted is False:
-            self.soln, self.residual, self.cond, self.linfit_result = (
-                leastsqnrm.matrix_operations(image, self.fittingmodel, dqm=dqm)
-            )
-        else:
+        if self.weighted:
```
Comment (braingram) [inline jwst/ami/lg_model.py:384]: I changed this check as it didn't quite match the log message about the kwarg in `FringeFitter`. The log message reported `Poisson` if `weighted is True` whereas this code tests `weighted is False`. If someone were to provide a non-bool value (1 for instance), the log would report equal weighting but poisson weighting would be used.

File: jwst/ami/tests/test_ami_leastsqnrm.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import numpy as np
+
+from jwst.ami import leastsqnrm
+
+
+def test_weighted_operations():
+    img = np.arange(4, dtype='f4').reshape((2, 2))
```
Comment (braingram) [inline jwst/ami/tests/test_ami_leastsqnrm.py]: @rcooper295 any suggestion for improving these fake values? At the moment this test passes but produces warnings:
```
jwst/ami/tests/test_ami_leastsqnrm.py::test_weighted_operations
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/leastsqnrm.py:500: RuntimeWarning: divide by zero encountered in divide
    weights = np.where(flatimg <= 1.0, 0.0, 1.0 / np.sqrt(variance))  # anand 2022 Jan

jwst/ami/tests/test_ami_leastsqnrm.py::test_weighted_operations
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/leastsqnrm.py:516: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
  To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
    x, rss, rank, singvals = np.linalg.lstsq(Aw, bw)
```
The first one looks like a result of the fake values. The second looks like `lstsq` should receive another argument.

Comment (rcooper295) [reply jwst/ami/tests/test_ami_leastsqnrm.py]: I think passing `rcond=None` to `lstsq` should be fine. Not sure how best to fix the first warning -- sounds like variance is 0 but I'm not sure why that's the case given your input "data"? When I print the variance it's [0. 1. 2.]

Comment (braingram) [reply jwst/ami/tests/test_ami_leastsqnrm.py]: Does this indicate a bug somewhere? Are you able to look into it? Your familiarity with the code will certainly help.

Comment (rcooper295) [reply jwst/ami/tests/test_ami_leastsqnrm.py]: I think it's just due to the zero in the test data. If I do `img = np.arange(1,5, dtype='f4').reshape((2, 2))` I don't get that warning anymore.

**Comment by braingram** [comment]
I made the test modification you suggested https://github.com/rcooper295/jwst/pull/7#discussion_r1523833961 (Thanks!)
and that test runs without warnings. The regtests still pass (tested locally).

I should note that the regtests show some warnings:
```
jwst/regtest/test_niriss_ami3.py::test_niriss_ami3_exp[012-ami-oi]
jwst/regtest/test_niriss_ami3.py::test_niriss_ami3_exp[012-ami-oi]
  /Users/bgraham/.pyenv/versions/3.10.6/envs/oifits/lib/python3.10/site-packages/numpy/lib/nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
    return np.nanmean(a, axis, out=out, keepdims=keepdims)
```
and the other unit tests also show warnings:
```
jwst/ami/tests/test_ami_interface.py::test_ami_analyze_step
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/bp_fix.py:323: RuntimeWarning: invalid value encountered in divide
    cvis /= np.max(cvis)

jwst/ami/tests/test_ami_interface.py::test_ami_analyze_step
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/bp_fix.py:186: RuntimeWarning: invalid value encountered in divide
    log.info("         %.3f" % np.max(diff_data / np.median(diff_data)))
```
Do any of these look suspicious? Do they seem easy to fix?

**Comment by rcooper295** [comment]
I think those warnings are again just due to the occasional zero in the test data or a clump of bad pixels marked as NaNs. If it's allowed to ignore them I think that's probably safe enough.

**Comment by braingram** [comment]
Thanks for the suggestion. I tried setting all the test data to `0.1` (in addition to the 2 slices set to `1.0`) in `example_model` and I see the same warnings for the unit test.

I'm not quite sure where the `nanmean` warning is coming from (it's unused in the ami submodule) and the other two as you noted seem likely caused by the unrepresentative `example_model`. Adding ignore filters for these would require finding where they're coming from (so we don't ignore all similar warnings which might indicate issues in other parts of the code). We could also leave them as-is if they seem innocuous although there is an effort underway to get a better handle of the tens of thousands of warnings the pipeline test runs current produce.

**Comment by rcooper295** [comment]
Hmm. I'm not sure what the best path forward is towards figuring out exactly why they're occurring and how to fix it, but in the interest of time if it's acceptable to let these warnings join the tens of thousands of other warnings produced by the pipeline tests, can we do that? 😬

**Comment by braingram** [comment]
I'm not sure. The reviewer(s) may want these filtered or the code adjusted to avoid the warning (so not dividing by `np.max(cvis)` if the max is 0). The unit test errors make me worry that the example model is unrepresentative or that the code will produce warnings for real data (which will concern users as `RuntimeWarning` will not be filtered in pipeline runs). The warning in the regression test is more worriesome as it is presumably representative of real data.

If you think it's worth a shot feel free to merge this PR (if there are no other issues) and if https://github.com/rcooper295/jwst/pull/7#discussion_r1523500513 looks acceptable. If you'd rather take more time to try and sort out the warnings let me know how I can help. I've tried a few modifications to `example_model` but at this point I'm just guessing as I'm not sure what would be a representative input.

**Comment by rcooper295** [comment]
I think having the mock NRM reference file contain some ones might help with some of the errors showing up in `bp_fix`, maybe just a blob of ones in the middle? e.g:
```
size = 1024
arr = np.zeros((size, size), dtype='f4')
middle = size//2
halfsize=50
arr[middle-halfsize:middle+halfsize,middle-halfsize:middle+halfsize] = 1
m.nrm = arr
```
I haven't tried this modification but from my understanding of how the BP fix works I think that might help. The regtest warning I'm a little more unsure of -- I am taking a nanmedian here: https://github.com/rcooper295/jwst/blob/amianalyze/jwst/ami/ami_analyze.py#L153  and there may be a place in the test data where that is all nans, which I think would give that empty slice warning. But it's nanmedian rather than nanmean?

**Comment by braingram** [comment]
Thanks for the suggestion. Unfortunately that modification results in a test failure:
```
>       d0, d1 = (c[0] - int(c[0]), c[1] - int(c[1]))
E       ValueError: cannot convert float NaN to integer

jwst/ami/hextransformee.py:114: ValueError
```
and more warnings:
```
jwst/ami/tests/test_ami_interface.py::test_ami_analyze_step
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/bp_fix.py:186: RuntimeWarning: invalid value encountered in divide
    log.info("         %.3f" % np.max(diff_data / np.median(diff_data)))

jwst/ami/tests/test_ami_interface.py::test_ami_analyze_step
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/utils.py:591: RuntimeWarning: invalid value encountered in divide
    cvmod = cvmod / cvmod.max()  # normalize to unity peak

jwst/ami/tests/test_ami_interface.py::test_ami_analyze_step
  /Users/bgraham/projects/230621_oifits/jwst/jwst/ami/utils.py:731: RuntimeWarning: Mean of empty slice.
    slopes = G[0] * tilt[0][newmaskh].mean(), G[1] * tilt[1][newmaskv].mean()

jwst/ami/tests/test_ami_interface.py::test_ami_analyze_step
  /Users/bgraham/.pyenv/versions/3.10.6/envs/oifits/lib/python3.10/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
    ret = ret.dtype.type(ret / rcount)
```
when running:
```
pytest jwst/ami/tests/test_ami_interface.py
```
Let me know if more information is helpful.

**Comment by rcooper295** [comment]
I spent some time experimenting with this to see if I could figure out what was causing those warnings and fix them, but I haven't been able to yet. If it's ok with you I think I'll go ahead and merge this so we can move forward with the main PR, and I'll note this down as a fix needed for a future version.

**Pull request merged**